core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -223,34 +219,29 @@ public ShardId shardId(ClusterState clusterState, String index, String id, @Null
     }
 
     static int generateShardId(IndexMetaData indexMetaData, @Nullable String id, @Nullable String routing) {
-        if (indexMetaData.isRoutingPartitionedIndex()) {
-            if (routing == null) {
-                throw new IllegalArgumentException("expected a routing value on partitioned index [" + indexMetaData.getIndex().getName() + "]");
-            }
+        final String effectiveRouting;
+        final int partitionOffset;
 
-            return generatePartitionedShardId(indexMetaData, routing, calculatePartitionOffset(indexMetaData, id));
+        if (routing == null) {
+            assert(!indexMetaData.isRoutingPartitionedIndex());
```

Comment:
we usually use `indexMetaData.isRoutingPartitionedIndex() == false` for easier readability. Also can you add some message to the assert ?